### PR TITLE
[SPARK-40750][SQL] Migrate type check failures of math expressions onto error classes

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
This pr replaces TypeCheckFailure by DataTypeMismatch in type checks in the math expressions, includes:
- hash.scala (HashExpression)
- mathExpressions.scala (RoundBase)

### Why are the changes needed?
Migration onto error classes unifies Spark SQL error messages.

### Does this PR introduce _any_ user-facing change?
Yes. The PR changes user-facing error messages.

### How was this patch tested?
- Add new UT
- Update existed UT
- Pass GA.

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -143,6 +143,11 @@
           "Offset expression <offset> must be a literal."
         ]
       },
+      "INVALID_ELEMENT_TYPE" : {
+        "message" : [
+          "Input to function <functionName> cannot contain elements of <elementType>. <message>"
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:148]: Let's don't pass `message`, and introduce more specific error class which includes the message.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json]: Done!

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -143,6 +143,11 @@
           "Offset expression <offset> must be a literal."
         ]
       },
+      "HASH_MAP_TYPE" : {
+        "message" : [
+          "Input to function <functionName> cannot contain elements of MAP. In Spark, same maps may have different hashcode, thus hash expressions are prohibited on MapType elements. To restore previous behavior set \"spark.sql.legacy.allowHashOnMapType\" to \"true\".",
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:148]:
```suggestion
          "Input to the function <functionName> cannot contain elements of the \"MAP\" type. In Spark, same maps may have different hashcode, thus hash expressions are prohibited on \"MAP\" elements. To restore previous behavior set \"spark.sql.legacy.allowHashOnMapType\" to \"true\".",
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 3]
```diff
@@ -268,15 +269,17 @@ abstract class HashExpression[E] extends Expression {
 
   override def checkInputDataTypes(): TypeCheckResult = {
     if (children.length < 1) {
-      TypeCheckResult.TypeCheckFailure(
-        s"input to function $prettyName requires at least one argument")
+      DataTypeMismatch(
+        errorSubClass = "WRONG_NUM_PARAMS",
+        messageParameters = Map(
+          "functionName" -> prettyName,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:275]: Please, quote the function id by `toSQLId()`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala [hunk 4]
```diff
@@ -268,15 +269,17 @@ abstract class HashExpression[E] extends Expression {
...
     } else if (children.exists(child => hasMapType(child.dataType)) &&
         !SQLConf.get.getConf(SQLConf.LEGACY_ALLOW_HASH_ON_MAPTYPE)) {
-      TypeCheckResult.TypeCheckFailure(
-        s"input to function $prettyName cannot contain elements of MapType. In Spark, same maps " +
-          "may have different hashcode, thus hash expressions are prohibited on MapType elements." +
-          s" To restore previous behavior set ${SQLConf.LEGACY_ALLOW_HASH_ON_MAPTYPE.key} " +
-          "to true.")
+      DataTypeMismatch(
+        errorSubClass = "HASH_MAP_TYPE",
+        messageParameters = Map("functionName" -> prettyName))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala:282]: Quote by `toSQLId()`

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (panbingkun) [reply core/src/main/resources/error/error-classes.json:148]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 4 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala (hunk 3 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala]: Done

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Pull request closed without merging**